navapbc/wic-participant-recertification-portal
# [PRP-166] Update CD for project with supporting changes

## Ticket

https://wicmtdp.atlassian.net/browse/PRP-166

## Changes
> What was added, updated, or removed in this PR.

- CI/CD:
  - Modify the CD github action to support independently deploying different apps within an environment
  - Modify the CD github action to become a reusable workflow and to use `aws-action` github actions to handle deploy
  - Add CD github actions for each of our three apps
  - Modify the participant portal CI and storybook CD github actions to only run if there are changes to themselves or the participant_portal dir
  - Modify Makefile and bin/publish-release.sh to support alternate a data model where there is one infra "app" but multiple project "apps"
  - Replace bin/deploy-release.sh with a bin/prepare-release.sh script to pass in envirnoment variables needed by `aws-action` github actions
  - Add back participant_portal/Makefile to support CD; set the target to the `prod` image
  - Fix bug in participant_portal/Dockerfile; `npm run css` must be done before `npm run build` because it requires participant_portal/app/styles/styles.css to exist
  - Use a unique build id as the image tag that links to the git commit and the github action workflow run
- Infra:
  - Add ECR repositories for staff portal and analytics
  - Output some more values to support CD
  - Fix bug in container-image-repository module that caused terraform to always force update
  - Use lifecycle `ignore_changes` on ECS task definition `container_definitions` so that Github Actions can manage deploys of new images
  - Change ECS task executor role policy to be a customer managed policy instead of an inline policy
- General:
  - Rename `participant_portal` dir to `participant`
  - Rename `staff_portal` dir to `staff`
- And more!

## Context for reviewers
> Testing instructions, background context, more in-depth details of the implementation, and anything else you'd like to call out or ask reviewers. Explain how the changes were verified.

This PR modifies our CD to support a slightly different data model than the [infra template](https://github.com/navapbc/template-infra/tree/3d7724a179da0d13b8a5d981890f428bec6f6005) currently supports. Namely, we want:
- multiple environments - this is the same; see infra/app/envs
- each environment has a single controlling terraform module - this is the same; see infra/app/envs/prod
- each environment has multiple applications (an application here is defined as something that needs a service, a load balancer, etc) that have shared resources (such as a database, IAM config, container registry, etc) - this is **not necessarily different**
- each environment should be able to independently deploy each application (i.e. we should be able to deploy the participant portal to the dev environment without having to also deploy the staff portal) - this is **different**; note that I think this only works is this reusable way because each of our apps is deployed the same way (e.g. build a docker image, publish the docker image, deploy the docker image)

Although we could organize the infra dir to have multiple "app" dirs (e.g. one for the participant portal, one for the staff portal, etc), this isn't actually what we want. We want to be able to share resources across our apps.

To accomplish this, I've made some changes. 
- I modified the CI/CD paradigm to support the concept of an "infra_app_name" (i.e. `app`) and an "app_name" (e.g. `participant`, `staff`). Each of these apps can be auto-deployed independently from one another
- I also turned the CD github action into a reusable workflow (it can still be used to manually deploy in the github web interface). 
- I also created a new wrapper github action that will check each of our app dirs to see if any files have been modified (e.g. added, renamed, deleted, changed) and deploy the app if there has been a change.
- I renamed the `participant_portal` dir to `participant` and the `staff_portal` dir to `staff` in order to support all the ids being one word (I was running into issues with max character limit on AWS resource names)

I ended up using the `lifecycle` `ignore_changed` policy on the Fargate ECS taske definition's `container_definitions`. This wasn't ideal, but https://github.com/hashicorp/terraform-provider-aws/issues/632 shows that there isn't really a better way to handle this at the moment. 

In addition, I've switched from using a custom deploy release script to following the [Github Actions recommended approach](https://docs.github.com/en/actions/deployment/deploying-to-your-cloud-provider/deploying-to-amazon-elastic-container-service), which uses:
- https://github.com/aws-actions/amazon-ecs-render-task-definition
- https://github.com/aws-actions/amazon-ecs-deploy-task-definition

### Feedback
It might be possible to DRY up the cd-on-push.yml, but I didn't want to keep pursuing this further. This can be saved for a future refactoring exercise. However, if you, the reviewer, know how to do this in Github Action syntax, please let me know!

## Testing
> Screenshots, GIF demos, code examples or output to help show the changes working as expected. ProTip: you can drag and drop or paste images into this textbox.

To verify the CD github action:
1. Go to https://github.com/navapbc/wic-participant-recertification-portal/actions/workflows/cd.yml
2. Select "Run workflow" and choose this branch (PRP-166-cd-connect-the-infrastructure). Leave the default settings. 
<img width="289" alt="Screen Shot 2023-03-15 at 1 58 53 AM" src="https://user-images.githubusercontent.com/67701/225258486-a39649c0-fd58-4307-9305-71bc06e368d4.png">

To verify the deploy on push github action:
1. [This workflow run](https://github.com/navapbc/wic-participant-recertification-portal/actions/runs/4424480309) demonstrates that it works when a change has been made to a file in a watched directory (in this case the participant_portal dir). Note that the other apps (staff_portal and analytics) did not trigger a deploy because no files were changed in those directories. 
<img width="1057" alt="Screen Shot 2023-03-15 at 2 14 14 AM" src="https://user-images.githubusercontent.com/67701/225262379-1551492d-51b2-4c4a-94d5-7c4d5e32ed88.png">

2. [This other workflow](https://github.com/navapbc/wic-participant-recertification-portal/actions/runs/4383341536) demonstrates that it works when a file is deleted in a watched directory.

To verify that the bugs in the participant_portal Makefile and Dockerfiles have been fixed (these are also tested in the above workflow runs, but here are the manual testing instructions):
1. Run `cd participant_portal && docker compose build`. You should get no errors.
2. Run `make release-build APP_NAME=participant_portal`. You should get no errors.
